README
:[diStorm64}: (v1.7.30)
The ultimate disassembler library (for AMD64, X86-64).

diStorm is licensed under the BSD license.

For documentation and MUCH MORE please refer to doc/distorm.html.


Directory Listing:
-+----------------
doc/ - Contains documentation files.
build/win32/ - Solution file for VS03.
build/linux/ - Make file and installation script for Python.
build/mac/ - Make file and installation script for Python2.3 (works for Tiger).
src/ - Full source code.
linuxproj/ - Raw file disassembler, sample project, shows how to use the C Library.
win32proj/ - Ditto, but for Win32.
ddkproj/ - A Windows kernel driver source that disassembles the beginning of KeBugCheck as an example.

distorm.h - Header file for diStorm's C library (Used by win32proj and linuxproj).
config.h - Compiler specific macros definitions (Endianity support too).

NOTE - See config.h and distorm.h for library configuration (such as exports, size of integers, etc).

Win32 Compilation:
-+----------------
The distorm.sln file contains two projects: cdistorm and pydistorm.
There are three configurations: clib, py24 and py23.
'clib' configuration will create a distorm.lib file for win32proj used with the
header file (distorm.h).
'py25/24/23' configuration will create distorm.pyd in the corresponding directory,
which should be copied into the Python\Lib\site-packages.
You might want to change the directories of the Python library/include directories
which are set in the configurations.
There is also a 'dyndistorm' project which will create a distorm.dll file which you can use along with 'distorm.h'.

Linux Compilation:
-+----------------
The makefile will compile diStorm for clib and py24.
'make clib' will create a static archive library named 'distorm64.a'
which is used by linuxproj.
'make py' will create a shared object file, named 'libdistorm64.so'
which is installed with the instpython.sh script (have to be root for that).

Mac Compilation:
-+--------------
The makefile will compile diStorm for clib and py23.
You can easily edit it to compile diStorm for py2x.
See the 'Linux Compilation' above for how to use the makefile.
You can also use the same sample of linux project.

DDK Compilation (Ring0):
-+----------------------
diStorm can also be compiled as a kernel driver,
there's a sample bundled under ddkproj.
Follow accurately all the steps in the ddkproj/readme file to accomplish this task.

For updates please visit http://ragestorm.net/distorm/
Note that there are currently not-known bugs.
Please contact me for any bugs that are found in the source code
by reporting to me, so people could enjoy a better product!

If you wish to contact me, use this email: arkon@ragestorm.net
Gil Dabah
May 2007


CHANGES
Copyright (C) 2003-2007 Gil Dabah, http://ragestorm.net/distorm/

History:
-+------
FPU(387) parsing added Feb 2004.
Prefixes parsing added July 2004.
SIB parsing added Dec 2004.
Rep/lock/xlat explicit-operand forms added Dec 2004.
Wrapped in Python Jan 2005.
Using static instructions tables Jan 2005.
MMX parsing added Feb 2005.
SSE parsing added Feb 2005.
SSE2 parsing added March 2005.
SSE3 parsing added March 2005.
Pseudo Opcodes parsing added March 2005.
3DNow! parsing added March 2005.
Global Optimization April 2005.
AMD64 added July 2005.
Linux Port November 2005.
Strings Optimization November 2005.
New API December 2005.
LGPL December 2005.
Doc + Sources uploaded Januar 2006.
64bit Offsets are now supported January 2006.
diStorm is now licensed under the BSD license January 2006.
1.2.8 diStorm can be compiled for Big Endian systems May 2006.
1.3.9: SSE4 (later known as SSSE3) parsing added May 2006.
1.3.9: VMX instructions set is now supported.
1.4.10: Instructions' flags were changed, engine supports instsruction set class type. May 2006.
1.4.10: All comments are now C style, source code is now 64bits compiler compatible.
1.5.10: Operand size of Memory Indirection is now determined inside operands.c and not instructions.c. May 2006.
1.5.15 SVM Instructions set is added Aug 2006.
Python2.5 is compiled too. Sep 2006.
1.5.17 Formerly unofficial SSE4 is now officially called SSSE3.
diStorm has a sample code for Ring0 using the DDK. Jan 2007.
1.6.18 Instructions table are now more optimized for size. Feb 2007.
1.6.21 Added a new Mac makefile to compile diStorm on a Mac. Apr 2007.
1.7.22 Real SSE4 parsing added April 2007.
1.7.26 New AMD instructions, support of different compilers.

Fixed bugs:
-+---------
2/11/05 1.0.1 Call instruction wasn't promoted to 64 bits automatically.
16/11/05 1.0.2 Mov MOFF instruction used RAX in 32 bits, and wasnt dropped in 16 bits.
17/01/06 1.1.3 Some special instructions were ignoring the REX.B flag.
27/01/06 1.1.4 Support for 64bit offsets was added.
03/02/06 1.1.5 PUSH, JMP and CMP instructions were slightly fixed.
18/03/06 1.1.6 MOVZX/MOVSX and RETF operands' types were changed.
20/03/06 1.1.7 A bug treating mandatory prefixes in 64bits was fixed.
18/04/06 1.1.8 LOOPxx instructions are now categorized as Natives.
18/05/06 1.2.9 A few FPU instructions weren't set correctly in IGEN.
28/05/06 1.4.10 Push/Pop instructions with ModR/M fields are now fixed.
02/06/06 1.5.11 Fixed a new bug in operands.c, where OT_IMM_FULL was broken in 64bits.
03/06/06 1.5.12 Invalid memory access occurred in 64bit environments only is now fixed.
09/06/06 1.5.13 SGDT, SIDT, LGDT, LIDT, JMP FAR, CALL FAR, SMSW, LMSW, LDS, LES, LFS, LGS, LSS and other instructions which have special memory operands' types were fully corrected. And a bug in locate_raw_inst when using mingw environment.
01/07/06 1.5.14 REX prefix with BASE bit isn't dropped always, and string instructions are now promoted to 64bits only with REX prefix.
08/07/06 1.5.14 The sample projects of both Linux and Windows have been updated.
02/10/06 1.5.16 MOVSXD wasn't decoded properly and it now supports MOVZXD.
08/01/07 1.5.17 Removed redundant #include's to standard libraries.
11/02/07 1.5.18 Fixed a bug in OT_RM32_64 decoding.
28/02/07 1.6.19 All string/IO instructions support both REP and REPNZ.
03/03/07 1.6.20 3DNow! instructions weren't decoded (produced in 1.6.19).
17/04/07 1.6.21 MOV-Offset instructions weren't decoded correctly in 64bits.
23/04/07 1.6.22 The mnemonic of MOVD/MOVQ wasn't changed according to operand size.
12/05/07 1.7.23 Added the instructions MOVSLDUP and INVLPGA and fixed SVM instructions.
03/07/07 1.7.24 Added multi-byte NOP instruction.
19/07/07 1.7.25 Fixed a memory leak in Python extension module.
07/12/07 1.7.27 Fixed various instructions.
09/02/08 1.7.28 NOP and XCHG are treated correctly in 64bits.
07/03/08 1.7.29 LEA now ignores segment override prefixes.
17/09/08 1.7.30 DLL project added to the Win32 solution and fixed DDK compilation.


win32proj/main.cpp
// diStorm64 library sample
// http://ragestorm.net/distorm/
// Arkon, Stefan, 2005


#include <windows.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "../distorm.h"

// Link the library into our project.
#pragma comment(lib, "../distorm.lib")

// The number of the array of instructions the decoder function will use to return the disassembled instructions.
// Play with this value for performance...
#define MAX_INSTRUCTIONS (1000)

int main(int argc, char **argv)
{
	// Version of used compiled library.
	unsigned long dver = 0;
	// Holds the result of the decoding.
	_DecodeResult res;
	// Decoded instruction information.
	_DecodedInst decodedInstructions[MAX_INSTRUCTIONS];
	// next is used for instruction's offset synchronization.
	// decodedInstructionsCount holds the count of filled instructions' array by the decoder.
	unsigned int decodedInstructionsCount = 0, i, next;

	// Default decoding mode is 32 bits, could be set by command line.
	_DecodeType dt = Decode32Bits;

	// Default offset for buffer is 0, could be set in command line.
	_OffsetType offset = 0;
	char* errch = NULL;

	// Index to file name in argv.
	int param = 1;

	// Handling file.
	HANDLE file;
	DWORD filesize, bytesread;

	// Buffer to disassemble.
	unsigned char *buf, *buf2;

	// Disassembler version.
	dver = distorm_version();
	printf("diStorm version: %d.%d.%d\n", (dver >> 16), ((dver) >> 8) & 0xff, dver & 0xff);

	// Check params.
	if (argc < 2 || argc > 4) {
		printf("Usage: disasm.exe [-b16] [-b64] filename [memory offset]\r\nRaw disassembler output.\r\nMemory offset is origin of binary file in memory (address in hex).\r\nDefault decoding mode is -b32.\r\nexample:   disasm -b16 demo.com 789a\r\n");
		return -1;
	}

	if (strncmp(argv[param], "-b16", 4) == 0) {
		dt = Decode16Bits;
		param++;
	} else if (strncmp(argv[param], "-b64", 4) == 0) {
		dt = Decode64Bits;
		param++;
	} else if (*argv[param] == '-') {
		printf("Decoding mode size isn't specified!");
		return -1;
	} else if (argc == 4) {
		printf("Too many parameters are set.");
		return -1;
	}
	if (param >= argc) {
		printf("Filename is missing.");
		return -1;
	}
	if (param + 1 == argc-1) { // extra param?
#ifdef SUPPORT_64BIT_OFFSET
		offset = _strtoui64(argv[param + 1], &errch, 16);
#else
		offset = strtoul(argv[param + 1], &errch, 16);
#endif
		if (*errch != '\0') {
			printf("Offset couldn't be converted.");
			return -1;
		}
	}

	file = CreateFile(argv[param], GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
	if (file == INVALID_HANDLE_VALUE) { 
		printf("Could not open file %s (error %d)\n", argv[param], GetLastError());
		return -2;
	}

	if ((filesize = GetFileSize(file, NULL)) < 0) {
		printf("Error getting filesize (error %d)\n", GetLastError());
		CloseHandle(file);
		return -3;
	}

	// We read the whole file into memory in order to make life easier,
	// otherwise we would have to synchronize the code buffer as well (so instructions won't be split).
	buf2 = buf = (unsigned char*)malloc(filesize);
	if (!ReadFile(file, buf, filesize, &bytesread, NULL)) {
		printf("Error reading file (error %d)\n", GetLastError());
		CloseHandle(file);
		free(buf);
		return -3;
	}

	if (filesize != bytesread) {
		printf("Internal read-error in system\n");
		CloseHandle(file);
		free(buf);
		return -3;
	}

	CloseHandle(file);

	printf("bits: %d\nfilename: %s\norigin: ", dt == Decode16Bits ? 16 : dt == Decode32Bits ? 32 : 64, argv[param]);
#ifdef SUPPORT_64BIT_OFFSET
	if (dt != Decode64Bits) printf("%08I64x\n", offset);
	else printf("%016I64x\n", offset);
#else
	printf("%08x\n", offset);
#endif

	// Decode the buffer at given offset (virtual address).
	while (1) {
		// If you get an unresolved external symbol linker error for the following line,
		// change the SUPPORT_64BIT_OFFSET in distorm.h.
		res = distorm_decode(offset, (const unsigned char*)buf, filesize, dt, decodedInstructions, MAX_INSTRUCTIONS, &decodedInstructionsCount);
		if (res == DECRES_INPUTERR) {
			// Null buffer? Decode type not 16/32/64?
			printf("Input error, halting!");
			free(buf2);
			return -4;
		}

		for (i = 0; i < decodedInstructionsCount; i++) {
#ifdef SUPPORT_64BIT_OFFSET
			printf("%0*I64x (%02d) %-24s %s%s%s\r\n", dt != Decode64Bits ? 8 : 16, decodedInstructions[i].offset, decodedInstructions[i].size, (char*)decodedInstructions[i].instructionHex.p, (char*)decodedInstructions[i].mnemonic.p, decodedInstructions[i].operands.length != 0 ? " " : "", (char*)decodedInstructions[i].operands.p);
#else
			printf("%08x (%02d) %-24s %s%s%s\r\n", decodedInstructions[i].offset, decodedInstructions[i].size, (char*)decodedInstructions[i].instructionHex.p, (char*)decodedInstructions[i].mnemonic.p, decodedInstructions[i].operands.length != 0 ? " " : "", (char*)decodedInstructions[i].operands.p);
#endif
		}

		if (res == DECRES_SUCCESS) break; // All instructions were decoded.
		else if (decodedInstructionsCount == 0) break;

		// Synchronize:
		next = (unsigned long)(decodedInstructions[decodedInstructionsCount-1].offset - offset);
		next += decodedInstructions[decodedInstructionsCount-1].size;
		// Advance ptr and recalc offset.
		buf += next;
		filesize -= next;
		offset += next;
	}

	// Release buffer
	free(buf2);

	return 0;
}


linuxproj/main.c
// diStorm64 library sample
// http://ragestorm.net/distorm/
// Arkon, Stefan, 2005
// Mikhail, 2006
// JvW, 2007

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <errno.h>
#include <sys/stat.h>

// For the compilers who don't have sysexits.h, which is not an ISO/ANSI include!
#define EX_OK           0
#define EX_USAGE       64
#define EX_DATAERR     65
#define EX_NOINPUT     66
#define EX_NOUSER      67
#define EX_NOHOST      68
#define EX_UNAVAILABLE 69
#define EX_SOFTWARE    70
#define EX_OSERR       71
#define EX_OSFILE      72
#define EX_CANTCREAT   73
#define EX_IOERR       74
#define EX_TEMPFAIL    75
#define EX_PROTOCOL    76
#define EX_NOPERM      77
#define EX_CONFIG      78

#include "../distorm.h"

// The number of the array of instructions the decoder function will use to return the disassembled instructions.
// Play with this value for performance...
#define MAX_INSTRUCTIONS (1000)

int main(int argc, char **argv)
{
	// Version of used compiled library.
	unsigned int dver = 0;
	// Holds the result of the decoding.
	_DecodeResult res;
	// Decoded instruction information.
	_DecodedInst decodedInstructions[MAX_INSTRUCTIONS];
	// next is used for instruction's offset synchronization.
	// decodedInstructionsCount holds the count of filled instructions' array by the decoder.
	unsigned int decodedInstructionsCount = 0, i, next;

	// Default decoding mode is 32 bits, could be set by command line.
	_DecodeType dt = Decode32Bits;

	// Default offset for buffer is 0, could be set in command line.
	_OffsetType offset = 0;
	char* errch = NULL;

	// Index to file name in argv.
	int param = 1;

	// Handling file.
	FILE* f;
	unsigned long filesize = 0, bytesread = 0;
	struct stat st;

	// Buffer to disassemble.
	unsigned char *buf, *buf2;

	// Disassembler version.
	dver = distorm_version();
	printf("diStorm version: %u.%u.%u\n", (dver >> 16), ((dver) >> 8) & 0xff, dver & 0xff);

	// Check params.
	if (argc < 2 || argc > 4) {
		printf("Usage: ./disasm [-b16] [-b64] filename [memory offset]\r\nRaw disassembler output.\r\nMemory offset is origin of binary file in memory (address in hex).\r\nDefault decoding mode is -b32.\r\nexample:   disasm -b16 demo.com 789a\r\n");
		return EX_USAGE;
	}

	if (strncmp(argv[param], "-b16", 4) == 0) {
		dt = Decode16Bits;
		param++;
	} else if (strncmp(argv[param], "-b64", 4) == 0) {
		dt = Decode64Bits;
		param++;
	} else if (*argv[param] == '-') {
		fputs("Decoding mode size isn't specified!\n", stderr);
		return EX_USAGE;
	} else if (argc == 4) {
		fputs("Too many parameters are set.\n", stderr);
		return EX_USAGE;
	}
	if (param >= argc) {
		fputs("Filename is missing.\n", stderr);
		return EX_USAGE;
	}
	if (param + 1 == argc-1) { // extra param?
#ifdef SUPPORT_64BIT_OFFSET
		offset = strtoull(argv[param + 1], &errch, 16);
#else
		offset = strtoul(argv[param + 1], &errch, 16);
#endif
		if (*errch != '\0') {
			fprintf(stderr, "Offset `%s' couldn't be converted.\n", argv[param + 1]);
			return EX_USAGE;
		}
	}

	f = fopen(argv[param], "rb");
	if (f == NULL) { 
		perror(argv[param]);
		return EX_NOINPUT;
	}

	if (fstat(fileno(f), &st) != 0) {
		perror("fstat");
		fclose(f);
		return EX_NOINPUT;
	}
	filesize = st.st_size;

	// We read the whole file into memory in order to make life easier,
	// otherwise we would have to synchronize the code buffer as well (so instructions won't be split).
	buf2 = buf = malloc(filesize);
	if (buf == NULL) {
		perror("File too large.");
		fclose(f);
		return EX_UNAVAILABLE;
	}
	bytesread = fread(buf, 1, filesize, f);
	if (bytesread != filesize) {
		perror("Can't read file into memory.");
		free(buf);
		fclose(f);
		return EX_IOERR;
	}

	fclose(f);

	printf("bits: %d\nfilename: %s\norigin: ", dt == Decode16Bits ? 16 : dt == Decode32Bits ? 32 : 64, argv[param]);
#ifdef SUPPORT_64BIT_OFFSET
	if (dt != Decode64Bits) printf("%08llx\n", offset);
	else printf("%016llx\n", offset);
#else
	printf("%08x\n", offset);
#endif

	// Decode the buffer at given offset (virtual address).
	while (1) {
		// If you get an undefined reference linker error for the following line,
		// change the SUPPORT_64BIT_OFFSET in distorm.h.
		res = distorm_decode(offset, (const unsigned char*)buf, filesize, dt, decodedInstructions, MAX_INSTRUCTIONS, &decodedInstructionsCount);
		if (res == DECRES_INPUTERR) {
			// Null buffer? Decode type not 16/32/64?
			fputs("Input error, halting!\n", stderr);
			free(buf2);
			return EX_SOFTWARE;
		}

		for (i = 0; i < decodedInstructionsCount; i++)
#ifdef SUPPORT_64BIT_OFFSET
			printf("%0*llx (%02d) %-24s %s%s%s\r\n", dt != Decode64Bits ? 8 : 16, decodedInstructions[i].offset, decodedInstructions[i].size, (char*)decodedInstructions[i].instructionHex.p, (char*)decodedInstructions[i].mnemonic.p, decodedInstructions[i].operands.length != 0 ? " " : "", (char*)decodedInstructions[i].operands.p);
#else
		printf("%08x (%02d) %-24s %s%s%s\r\n", decodedInstructions[i].offset, decodedInstructions[i].size, (char*)decodedInstructions[i].instructionHex.p, (char*)decodedInstructions[i].mnemonic.p, decodedInstructions[i].operands.length != 0 ? " " : "", (char*)decodedInstructions[i].operands.p);
#endif

		if (res == DECRES_SUCCESS) break; // All instructions were decoded.
		else if (decodedInstructionsCount == 0) break;

		// Synchronize:
		next = (unsigned int)(decodedInstructions[decodedInstructionsCount-1].offset - offset);
		next += decodedInstructions[decodedInstructionsCount-1].size;
		// Advance ptr and recalc offset.
		buf += next;
		filesize -= next;
		offset += next;
	}

	// Release buffer
	free(buf2);

	return EX_OK;
}
